### ISSUE TITLE
Success Criterion 3.2.2 - On Input - Level A

### ISSUE THREAD

**Issue by JJdeGroot** [w3c/matf#42]
WCAG2ICT guidance: https://www.w3.org/TR/wcag2ict-22/#on-input

Share your thoughts for applying to mobile apps as a comment below.

**Comment by JJdeGroot** [comment]
Discussed in yesterday's meeting.

<details>
<summary>9 October 2024</summary>

[Source](https://www.w3.org/2024/10/09-matf-minutes.html#a774)

~~~text
JJ: (looking at WCAG2ICT text for 3.2.2 on input)
… reading intent
… looking at the MATF Github issue

Joe_Humbert: We also need to look at the definition of user interface component, check if we need to modify that
… what is it we think a control is?

JJ: same note about user interface element

JJ: agenda is finnished - any additional topics?

Jamie: what is the next SC?

JJ: (checking which is next - consistent navigation)

Joe: Should be staret a document for "proposed definitions" so we can work outside meetings?

ACTION: Mark Github issues related to Definitions with 'definition' tag
~~~
</details>

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>4 December 2024</summary>

[Source](https://www.w3.org/2024/12/04-matf-minutes.html#4423)

~~~text

3.2.2 On Input

JJ: labelled as small variation
… WCAG2ICT says applies as written
… we discussed what counts as User Interface Component

<quintinb> Are components different in mobile?

<quintinb> Seems legit to me

JJ: reading UIC definition in WCAG
… comparing with 3.2.1 On focus

Detlev: mentioning two types of focus, so this may need clarification

JJ: Reading the WCAG2ICT note for 3.2.1

Joe: Reviewing definitions UIC "users perceive as control" - but don't define control

<Karla> Agree, the example in the definition of "user interface component" is not pretty clear, maybe we can change it

Joe_Humbert: definition of contro needed or new example - example confusing

JJ: We had similar difficulties with definitions

Jamie: We would not be keeping it as written - the example should be more oriented at mobile app context
… element and control are nti the same, element is broader
… we vary from WCAG2ICT variations at least in definitions

ACTION: Clarify "user interface component" definition in mobile context

JJ: change of context definitionmay need update - UA not applicable - viewport definition different
… content is also fouced on web content

ACTION: Clarify "change of context" definition in mobile context (used in 3.2.1 and 3.2.2)

JJ: in 3.2.1 and 3.2.2 - only places where it appears
… user interface component used more broadly

Jon: a change of setting implies that the component changes the context - web issues around on input these context would have a value attached to it... does it apply to this context?

JJ: We would look at a subset of UIC - so it onl yapplies to those that change settings - nut all components have settings - a button would not fall under this

ACTION: take a closer look at what "Changing the setting" means in 3.2.2

JJ: furthe rcomments?

~~~
</details>

- ACTION: Clarify "user interface component" definition in mobile context
- ACTION: Clarify "change of context" definition in mobile context (used in 3.2.1 and 3.2.2)
- ACTION: take a closer look at what "Changing the setting" means in 3.2.2

**Comment by JJdeGroot** [comment]
Discussed in today’s meeting.

<details>
<summary>9 April 2025</summary>

[Source](https://www.w3.org/2025/04/09-matf-minutes.html#ea37)

~~~text
<pauljadam> Ok mobile techniques need to go here too https://www.w3.org/WAI/WCAG21/Techniques/

<Jamie> Did I miss a vote?

<Jon_Gibbins> pauljadam: I agree, eventually. What I was saying is that, until such time as we can work on understanding and techniques documents (currently beyond our group’s remit, I believe) linking to examples may be the best we can do for now.

JJ: Summarizes 3.2.2, including need to define user interface component and change of context.

JJ: Further definition also required for "changing the setting"

<JJ> https://w3c.github.io/matf/#success-criterion-3-2-2-on-input

<JJ> [w3c/matf#42](https://github.com/w3c/matf/issues/42)

<pauljadam> That is failing WCAG to auto submit the form unless they are warned of the behavior before hand.

<pauljadam> How do they know

<pauljadam> you must tell them

julianmka: Common pattern where a one-time passcode is automatically submitted after the user enters it. Seems to be commonly done and has sort of become expected behavior.

JJ: Right, do we need to inform users if it has become an expected pattern? Users seem to understand that the PIN screen will submit.

<Jon_Gibbins> I have seen this in autocomplete widgets, but mostly on web and I admit rarely on mobile. Sometimes, developers assume that moving focus on behalf of the user to search results is necessary, which is horrible when typing into an autocomplete.

<Joe_Humbert> on android the auto-submit is a user option

Tanya: We see forms changing content after selecting a checkbox and fail that.

JJ: On Android or iOS, when you enter the PIN and it is automatically submitted. Some people were confused when they had to press a button after entering.

<pauljadam> Just a simple text message above the input warning of the behavior is all that is required.

JJ: For Android and iOS, have users been informed of the behavior during onboarding?

<pauljadam> That would be a perfect mobile a11y technique :)

Potentially we could use a note if it's common behavior. It's up to auditors to decide if the user has been sufficiently informed.

JJ: We have one client who added a Submit button. But maybe you could have have some text on the PIN screen advising that the code will be automatically submitted.

<pauljadam> "common behavior" is not the same as "the user has been advised of the behavior before using the component."

<pauljadam> https://www.w3.org/WAI/WCAG21/Techniques/general/G13

<pauljadam> DOB split into 3 inputs is another somewhat common example

ACTION: Potentially add note about pincode screen auto-submit behavior?

<pauljadam> Usually these are web views and they would be text field inputs

<pauljadam> usually it will be text fields that are made to accept the passcode automatically like with auto fill

JJ: On iOS, there's also specific code to pick up the one-time passcode from messages or email, and it autofills the code and automatically submits the form.

<Jon_Gibbins> JJ +!

<JJ> https://github.com/w3c/matf/issues?q=is%3Aissue%20state%3Aopen%20label%3Adefinition

JJ: Mostly we need to focus on definitions for change of context and user interface component. I suggest that we propose specific definitions to the group and discuss them in our meetings.

<Jon_Gibbins> JJ +1

TF members, please add yourself as an assignee to the definition issues in Github.
~~~
</details>

- ACTION: Potentially add note about pincode screen auto-submit behavior?
- TF members, please add yourself as an assignee to the definition issues in Github.

**Comment by JJdeGroot** [comment]
[Discussed in MATF meeting on 25 February 2026](https://www.w3.org/2026/03/04-matf-minutes.html#4f36)

<details>
<summary>Minutes for 4 March 2026</summary>

~~~text
<JJ> [w3c/matf#42](https://github.com/w3c/matf/issues/42)

There is a lot of search functionality that looks like a textbox, but behaves like a button taking you to a search screen

<pauljadam> I've seen those bugs before quintinb

So you agree it's a violation?

(I think it is)

<pauljadam> yes

I really appreciate not feeling alone :cry-smile:

<pauljadam> but I think I failed it under On Focus

ah

<JJ> Poll: Is it sufficient for MATF to link to 3.2.2 On Input's Intent section for "Changing the setting"?

<pauljadam> +1

<quintinb> +1

<shoobe01> +1

<Tim> +1

<JatinV> +1

<tayef> +1

ACTION: For "changing the setting" 3.2.2 link to the Intent section the explains what it means

<JJ> Poll: For 3.2.2 On Input, do we want to add examples of common mobile user interaction patterns that will count for "user has been advised of the behavior"?

<pauljadam> that sounds like a technique

<pauljadam> 0

<Tim> +1

<JatinV> +1 OTP (one time passcode) is another example

<tayef> +1

<quintinb> +1+1

Soz

too keen

trying to vote for Joe

<shoobe01> +1

ACTION: 5x +1 and 1x 0 for adding common mobile user interaction patterns that count for user has been advised of the behavior"
~~~
</details>

**Poll:** Is it sufficient for MATF to link to 3.2.2 On Input's Intent section for "Changing the setting"?

**Result:** 6x +1 -> accepted

- ACTION: For "changing the setting" 3.2.2 link to the Intent section the explains what it means

**Poll:** For 3.2.2 On Input, do we want to add examples of common mobile user interaction patterns that will count for "user has been advised of the behavior"?

Result: 5x +1 and 1x 0 -> mostly positive

- ACTION: add common mobile user interaction patterns that count for: "user has been advised of the behavior"

**Comment by jha11y** [comment]
I will add a link to https://www.w3.org/WAI/WCAG22/Understanding/on-input#intent in a note section

I need examples of 

> common mobile user interaction patterns that count for: "user has been advised of the behavior"

to make the other requested updates

**Comment by TanyaVanWorkum** [comment]
[Discussed in MATF meeting on 1 July 2026](https://www.w3.org/2026/07/01-matf-minutes.html)

<details>
<summary>Minutes</summary>

~~~text
ol - On Input, taken up
<pauljadam> Understanding page for that SC says "This behavior of phone fields is described for the user at the beginning of the form."
<pauljadam>[ https://www.w3.org/WAI/WCAG22/Techniques/general/G13](https://www.w3.org/WAI/WCAG22/Techniques/general/G13)
<pauljadam> Examples
<pauljadam> A select dropdown at the top of a page includes options for German, French and Spanish. Instructions preceding the dropdown instruct the user that the language will be changed upon selecting an option.
<pauljadam> A 50-question online survey displays one question at a time, with each answer marked up as a toggle button. Instructions appear at the beginning of the survey that explain that users will be taken to the next question of the survey upon selecting an answer to each question.
Joe_Humbert I don't have examples off hand. User are advised of behaviour beforehand - I need examples from a mobile perspective
Joe_Humbert if you can give me examples I'll update the language
Joe_Humbert are those app patterns?
pauljadam one example is for date of birth - it automatically jumps across fields
Joe_Humbert are those specifi mobile app examples?
Joe_Humbert I don't see the point of adding generic examples
pauljadam I don't see how we need to tweak it for mobile
Joe_Humbert The request for examples was because people may not have mobile specific examples
pauljadam there may be confusion around context versus content - that they need to warn the user before. That's not really required.
Joe_Humbert If you look at change of context does have content change
pauljadam only if it changes the meaning of the page
pauljadam it's not well defined
Joe_Humbert That's a larger problem - it seems we don't need examples - I guess we can call it good
Joe_Humbert the intent section doesn't have an id for me to point to
<pauljadam> a new mobile / modern example could be the 2 factor authentication code where when you type all the values then it automatically submits I guess that would need a warning before hand right?
Detlev do we need to think of the case of things like radio inputs behaving like buttons - like controls that create an expectation and then a change of context happens, but for something like a radio and in a range of questions when an option is selected it changes the screen, but it's not ok if it's a radio (ok for a button)
pauljadam that sounds like an example in G.13 - when clicking an answer it goes to the next page
pauljadam if you need a modern example - 2FA is an example, when you type the last digit it auto-submits
I hate that pattern, pauljadam
(on a personal level)
<pauljadam> quintinb well it does save you time so if there is a warning and it makes things faster for the user then seems fine to me :)
I love it when it auto-copies from messages!
<pauljadam> Like when apple auto pastes in the code then it submits and you dont have to click the submit button
aww yeah!
<shoobe01> Yes, I'll take help with scope of discussion for #247. Not sure what all is available as options but happy to write once I get that.
<shoobe01> Maybe collaborate with someone else?
Yeah add me to it - I'll add some comments
~~~
</details>

**ACTIONS:** add a link to https://www.w3.org/WAI/WCAG22/Understanding/on-input#intent in a note section. 
**ACTIONS** do not add anything else, we have not concluded any common mobile user interaction patterns that count for: "user has been advised of the behavior"
